byron/ledger/executable-spec/test/Ledger/Relation/Properties.hs

```diff
@@ -0,0 +1,223 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ledger.Relation.Properties
+    (testRelation)
+  where
+
+import           Data.Bimap          (Bimap)
+import qualified Data.Bimap          as Bimap
+import           Data.Map.Strict     (Map)
+import qualified Data.Map.Strict     as Map
+import           Data.Set            (Set)
+import qualified Data.Set            as Set
+
+import           Hedgehog            (Gen, MonadTest, Property, PropertyT,
+                                      classify, forAll, property, withTests,
+                                      (===))
+
+import qualified Hedgehog.Gen        as Gen
+import qualified Hedgehog.Range      as Range
+
+import           Test.Tasty.Hedgehog
+
+import           Ledger.Core         hiding ((<|))
+
+import           Test.Tasty          (TestTree, testGroup)
+
+--------------------------------------------------------------------------------
+-- Properties on Relations
+--------------------------------------------------------------------------------
+
+-- | (dom r) ∩ s == dom (s ◁ r)
+--
+--  By restricting the domain of 'Relation r' to 's', the new 'Domain r'
+--  is an intersection of the original 'Domain r' with 's'
+propDomainRestrictionAndIntersection
+  :: (MonadTest m, Relation r, Ord (Domain r))
+  => Domain r -> Set (Domain r) -> r -> m ()
+propDomainRestrictionAndIntersection x s r
+  = Set.member x (dom r `Set.intersection` s)
+    ===
+    Set.member x (dom (s ◁ r))
+
+-- | (dom r ∩ s) ◁ r == s ◁ r
```

Comment:
I wonder whether `(s ∩ t) ◁ r ==( s ◁ r) ∩( t ◁ r)` and `r ∩( s ◁ r) == r ∩( s ◁ r)` . If so this property might be a consequence of the precedent equalities and the fact that   `dom r  ◁ r == r`.

Reply:
it turns out (s ∩ t) ◁ r ==( s ◁ r) U ( t ◁ r) is not a property (e.g. s = {}, t = {0}, r = {0 -> 0}))